tests/loader_phys_dev_inst_ext_tests.cpp

```diff
@@ -0,0 +1,4730 @@
+
+// These tests are all instance extension tests that touch physical devices.  This was
+// before the idea that physical device extensions were more appropriately found in the
+// list of device extensions.  Because of that, all these tests need to support devices
+// that don't support the extension and have a fallback path in the loader that needs
+// validation.
+
+class LoaderInstPhysDevExts : public ::testing::Test {
+   protected:
+    virtual void SetUp() {
+        env = std::unique_ptr<FrameworkEnvironment>(new FrameworkEnvironment());
+        env->add_icd(TestICDDetails(TEST_ICD_PATH_VERSION_2));
+    }
+
+    virtual void TearDown() { env.reset(); }
+    std::unique_ptr<FrameworkEnvironment> env;
+};
+
+// Fill in random but valid data into the device properties struct for the current physical device
+static void FillInRandomICDInfo(uint32_t& vendor_id, uint32_t& driver_vers) {
+    vendor_id = VK_MAKE_API_VERSION(0, rand() % 64, rand() % 255, rand() % 255);
+    driver_vers = VK_MAKE_API_VERSION(0, rand() % 64, rand() % 255, rand() % 255);
+}
+
+// Fill in random but valid data into the device properties struct for the current physical device
+static void FillInRandomDeviceProps(VkPhysicalDeviceProperties& props, uint32_t api_vers, uint32_t vendor, uint32_t driver_vers) {
+    props.apiVersion = api_vers;
+    props.driverVersion = driver_vers;
+    props.vendorID = vendor;
+    props.deviceID = (static_cast<uint32_t>(rand()) >> 4) + (static_cast<uint32_t>(rand()) << 2);
+    props.deviceType = static_cast<VkPhysicalDeviceType>(rand() % 5);
+    for (uint8_t idx = 0; idx < VK_UUID_SIZE; ++idx) {
+        props.pipelineCacheUUID[idx] = static_cast<uint8_t>(rand() % 255);
+    }
+}
+
+//
+// VK_KHR_get_physical_device_properties2
+//
+
+// Test vkGetPhysicalDeviceProperties2KHR where nothing supports it.
+TEST_F(LoaderInstPhysDevExts, PhysDevProps2KHRNoSupport) {
+    FrameworkEnvironment env{};
+    env.add_icd(TestICDDetails(TEST_ICD_PATH_VERSION_6));
+    env.get_test_icd(0).physical_devices.push_back({});
+
+    InstWrapper instance(env.vulkan_functions);
+    instance.CheckCreate();
+
+    auto GetPhysDevProps2 = reinterpret_cast<PFN_vkGetPhysicalDeviceProperties2KHR>(
+        instance.functions->vkGetInstanceProcAddr(instance, "vkGetPhysicalDeviceProperties2KHR"));
+    ASSERT_EQ(GetPhysDevProps2, nullptr);
+}
+
+// Test vkGetPhysicalDeviceProperties2KHR where instance supports it, but nothing else.
+TEST_F(LoaderInstPhysDevExts, PhysDevProps2KHRNoICDSupport) {
+    FrameworkEnvironment env{};
+    env.add_icd(TestICDDetails(TEST_ICD_PATH_VERSION_6));
+    env.get_test_icd(0).physical_devices.push_back({});
+
+    InstWrapper instance(env.vulkan_functions);
+    instance.create_info.add_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+    instance.CheckCreate(VK_ERROR_EXTENSION_NOT_PRESENT);
+
+    auto GetPhysDevProps2 = reinterpret_cast<PFN_vkGetPhysicalDeviceProperties2KHR>(
```

Comment:
I really should of added something to make getting function pointers muuuch easier.